proposals/20200909-userpsace-deployment.md

```diff
@@ -0,0 +1,163 @@
+
+
+## Workflow
+### Preparation (done by Falco maintainers)
+We prepare and publish a docker image containing Falco. The contents would be 
+the following:
+
+```
+/falco/falco - falco binary
+/falco/pdig - ptrace-based instrumentation
+/falco/* - falco configs? other binaries?
+```
+
+We also provide a file called `kilt.cfg` which looks like:
+
+```
+ build {
+    entrypoint: [ /falco/pdig ] ${original.entrypoint}
+    mount: [
+        {
+            image: "falco/falco:latest"
+            volumes: /falco
+            entrypoint: /falco/waitforever
+        }
+    ]
+	environment:
+		FALCO_METADATA: "image="${original.image}",name="${original.name}
+}
+runtime {
+    upload: [
+        {
+            url: "http://blah"
+            as: "/falco/falco"
+        }
+        {
+            file: "local/path/to/pdig"
+            as: "/falco/pdig"
+        }
+        {
+            payload: "base64payload" //or directly text
+            as: "/falco/falco.cfg"
+        }
+    ]
+    exec: [
+        /falco/pdig -p 1
+    ]
+}
+```
+
+### Cloud Formation User Workflow
+#### What the final user sees
+User downloads an installer done by us and executes it.
+User then adds the newly created macro as a transform to the Cloud Formation templates that wants monitored and runs
+ a deploy.
+ 
+This step can also be used to distribute upgrades for Falco and Kilt runtimes.
+
+#### Behind the scenes
+The installer downloads the `kilt` binary and sets it up as an
+ `AWS::Lambda`. A [CFN Macro](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html)
+ is registered with a name provided by the user (example: `FalcoMacro`).
+
+### Kilt Transformation
+
+#### What the final user sees
+User creates a new stack/upgrades an existing stack with a CFN template that includes `Transform: ${MacroName}`
+ at root level.
+
+#### Behind the scenes
+The kilt-cfn lambda is invoked on each stack creation/update referencing the
+ macro. It finds all fargate task definitions in the template and applies the 
+patches defined in `kilt.cfg` specified during macro installation.
+  
+### Runtime
+#### What the final user sees
+final user sees logs from Falco in cloudwatch
+
+#### Behind the scenes
+pdig is invoked. It forks and execs Falco and original command.
+
+
+## Design
+
+### kilt-cfn
+A aws lambda written in go that receives fragments of template to transform. 
+Does not need any permissions. Will execute instructions in the provided 
+`kilt.cfg` file. Can only interpret directives under `build`
+
+### kilt-k8s-webhook
+A webservice to be deployed to accept admission controller webhooks. 
+Applies only `build` directives. The service will run under k8s and will be 
+deployed via YAML.
+
+### kilt-k8s-exec
+A CLI that uses local kubectl to instrument containers. Will only interpret 
+`runtime` directives
+
+## To be decided
+* config format - is [HOCON](https://github.com/lightbend/config/blob/master/HOCON.md) fine?
```

Comment:
Oh, so the configuration format is undecided yet.

This sounds like a topic for Community Call :)

Reply:
Yeah, forgot to remove that.